llm-examples/falcon/falcon-7b-instruct-inference.py

```diff
@@ -0,0 +1,349 @@
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Inference
+# MAGIC The below snippets are adapted from [the model card of falcon-7b-instruct](https://huggingface.co/tiiuae/falcon-7b-instruct). The example in the model card should also work on Databricks with the same environment.
+
+# COMMAND ----------
+
+# Load model to text generation pipeline
+
+from transformers import AutoTokenizer, AutoModelForCausalLM
+import transformers
+import torch
+
+model = "tiiuae/falcon-7b-instruct"
+
+tokenizer = AutoTokenizer.from_pretrained(model)
+pipeline = transformers.pipeline(
+    "text-generation",
+    model=model,
+    tokenizer=tokenizer,
+    torch_dtype=torch.bfloat16,
+    trust_remote_code=True,
+    device_map="auto",
+    revision="9f16e66a0235c4ba24e321e3be86dd347a7911a0", # it is a suggested practice to pin the revision commit hash
+)
+
+# COMMAND ----------
+
+# Define prompt template, see why we need this format: http://fastml.com/how-to-train-your-own-chatgpt-alpaca-style-part-one/
+INSTRUCTION_KEY = "### Instruction:"
+RESPONSE_KEY = "### Response:"
+INTRO_BLURB = "Below is an instruction that describes a task. Write a response that appropriately completes the request."
+PROMPT_FOR_GENERATION_FORMAT = """{intro}
+{instruction_key}
+{instruction}
+{response_key}
+""".format(
+    intro=INTRO_BLURB,
+    instruction_key=INSTRUCTION_KEY,
+    instruction="{instruction}",
+    response_key=RESPONSE_KEY,
+)
+
+# COMMAND ----------
+
+# Define parameters to generate text
+def gen_text(prompt, use_template=False, **kwargs):
+    if use_template:
+      full_prompt = PROMPT_FOR_GENERATION_FORMAT.format(instruction=prompt)
+    else:
+      full_prompt = prompt
+    
+    # configure text generation kwargs
+    if "max_new_tokens" not in kwargs:
+        kwargs["max_new_tokens"] = 512
+    
+    kwargs.update({
+      'do_sample': True,
+      'use_cache': True,
+      'num_return_sequences': 1,
+      'pad_token_id': tokenizer.eos_token_id,
+      'eos_token_id': tokenizer.eos_token_id
+    })
+
+    output = pipeline(full_prompt, **kwargs)[0]['generated_text']
+    return output
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Inference on a single input
+
+# COMMAND ----------
+
+result = gen_text("What is a large language model?")
+print(result)
+
+# COMMAND ----------
+
+result = gen_text("What is a large language model?", temperature=0.5)
+print(result)
+
+# COMMAND ----------
+
+result = gen_text("What is a large language model?", max_new_tokens=20)
+print(result)
+
+# COMMAND ----------
+
+result = gen_text("What is a large language model?", use_template=True)
+print(result)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Batch inference
+
+# COMMAND ----------
+
+# Required tokenizer setting for batch inference
+pipeline.tokenizer.pad_token_id = tokenizer.eos_token_id
+pipeline.tokenizer.padding_side = 'left'
+
+# COMMAND ----------
+
+# Define parameters to generate text
+def batch_gen_text(prompts, use_template=False, **kwargs):
```

Comment:
what's the difference between `gen_text` and `batch_gen_text `? can you just reuse the functionality say 

`def gen_text(prompt, use_template=False,  batch_size=1, **kwargs):`

Reply:
Also the input / output is a string vs. list of strings, feels simpler for `gen_text` to just treat the input as a string.

Reply:
Could we define one function to handle both cases?

Reply:
+1 for defining one function. 

The notebook we publish will implicitly represent the "best practice" DataBricks would  recommend. Therefore, we would like to avoid common pitfalls like code duplication and embrace good practices like detailed documentation :)

Reply:
actually, do you need `gen_text` at all? I feel the single batched version is good enough, non-batch version is really just special case (with `prompts` array len to be just 1)

Reply:
Got it! Now merged the functions into a single one.